hpx-sys/src/lib.rs

```diff
@@ -39,9 +40,45 @@ pub mod ffi {
 // Wrapper for the above Bindings.
 // reffer to tests to understand how to use them. [NOTE: Not all bindings have wrapper.]
 // ================================================================================================
-use std::ffi::CString;
+use std::env::Args;
+use std::ffi::{CString, CStr};
 use std::os::raw::c_char;
 
+static mut FUNC_HOLDER: Option<fn (Vec<String>) -> i32> = None;
+
+// Convert arguments from *mut *mut c_char to Vec<String>
+// and call the saved Rust version of the function.
+fn c_init(argc: i32, argv: *mut *mut c_char) -> i32 {
+    let mut vec_args: Vec<String> = Vec::new();
+
+    for i in 0..argc {
+        // SAFETY: We get the argv from a conversion from Vec<String>.
+        // The conversion, which is safe, perserves the arguments.
+        // Therefore, no null pointers.
+
+        let c_str: &CStr = unsafe { CStr::from_ptr(*argv.offset(i as isize)) };
+        let string = c_str.to_string_lossy().into_owned();
+        vec_args.push(string.to_string());
+    }
+
+    unsafe { FUNC_HOLDER.unwrap()(vec_args) }
+}
+
+pub fn init(func: fn(Vec<String>) -> i32, func_args: Vec<String>) -> i32
+{
+    unsafe { FUNC_HOLDER = Some(func) };
+    let str_args: Vec<&str> = func_args
+        .iter()
+        .map(|s| s.as_str())
+        .collect::<Vec<&str>>();
+    let (argc, rust_argv) = create_c_args(&str_args);
+    let argv = rust_argv.into_raw_parts().0;
+
+    unsafe {
+        return self::ffi::init(c_init, argc, argv);
+    }
+}
```

Comment:
remove the `static mut` variable and replace it with `boxed closure` to store callback. use `lifetime` with closure because it needs to live until explicitly freed.

Reply:
Do we really need the `FUNC_HOLDER`? Can we not be a bit more barbaric and construct `vec_args` within here? This makes the `FUNC_HOLDER` go away at no cost.

The need to have a static variable global to this file is more of a burden to verify with the best practices and I intend to just go around it until I know better.

Reply:
You cannot put the ``c_init`` function inside of the rust ``init`` function because it captures the dynamic environment (the Vec<String> function pointer). You cannot use a closure because that's not what the FFI allows. So the only way to do this is as follows (in pseudocode):

```
fn some_rust_callback(argc: i32, argv: *mut *mut c_char) -> i32 {
let args = c_args_to_vec(argc, argv);
return rust_func(args); // this is the problem! how do we get this function in scope?
}

fn rust_init(rust_func: fn(Vec<String>) -> i32, args: Vec<String>) -> i32 {
let (argc, argv) = vec_to_c_args(args);
return self::ffi::init(some_rust_callback, argc, argv);
}
```